pep-0497-externals.txt

```diff
@@ -0,0 +1,252 @@
+PEP: 497
+Title: Enhanced handling of external requirements
+Version: $Revision$
+Last-Modified: $Date$
+Author: Tennessee Leeuwenburg <tleeuwenburg@gmail.com>,
+
+Status: Draft
+Type: Standards Track
+Python-Version 3.6
+Content-Type: text/x-rst
+Created: 03-Aug-2015
+Post-History: 03-Aug-2015
+
+
+Abstract
+========
+
+The canonical method for installing packages is to use pip, whether that is
+referencing PyPI or a local repository. A limitation of the current install
+ecosystem is the relationship with operating-system level dependencies. This
+can result in compilation failures or execution failures. Package maintainers
+may wish to specify the requirements to avoid this issue, and users may
+appreciate earlier stopping and more informative messages about their system.
+
+The responsibility of pip is to stop early and provide useful information about
+missing and not automatically installable dependecies. Ideally it would provide
+a useful user message about how to obtain the required dependencies. The
+responsibility for actual installation lies with the user at this time.
+
+This PEP specifies an extension to the requirements sections of setup.py
+which will allow authors to express these requirements cleanly. In the
+implementation section, a discussion of how this is used to solved problems
+is given.
+
+Here is an example of the proposed syntax, where both operating system and build
+target specifiers are included.
+
+::
+
+    setup(
+        name='uses_scipy_with_atlas_extension'
+        install_requires='yaml, scipy[atlas]',
+        setup_requires='pyfortran',
+        external={
+          setup=[
+            'bin!fortran:dist_name=='ubuntu',
+            'include!kernel.h',
+            'include!liblas.h',
+          ],
+          install=[
+              'lib!liblas1.so'
+          ],
+       }
+    )
+
+It is believe this syntax provides one obvious way to do things which does not
+multiply the complexity or parts beyond those necessary to achieve the goal.
+
+In this example, the user will be presented with a useful install message and
+a process that exist early, as opposed to relying on the compiler to fail
+and provide an error message that may not be readily understood by most users
+
+The mechanism of referencing failed dependencies such as 'kernel.h' to useful
+user messages regarding the actual operating system packages that are relevant
+is discussed in the implementation section, along with the rationale for
+handling that referencing outside of setup.py.
+
+Rationale
+=========
+
+Different environment context have implications for the building, testing,
+installing and execution of software. There are two sides to this coin. First,
+in some environments, additional python packages are needed. Secondly,
+it is useful to specify, e.g., the apt or yum package names which need to be
+in place prior to continuing. The syntax proposed here allows for a readable
+and consise expression of these requirements.
```

Comment:
The domain of the two things - Python package names and APT package names are quite different - and we've settled (for now at least) on artifacts rather than the abstract APT package names. So I think this should say something like

'in some environments additional Python packages are needed. Secondly, in some environments non-Python package resources such as header files, libraries, scripts and so on are needed. The syntax proposed here allows describing both sorts of requirements in a single syntax.